src/main/java/org/truffleruby/language/LazyRubyNode.java

```diff
@@ -63,11 +63,13 @@ public RubyNode resolve() {
                 Log.LOGGER.info(() -> "lazy translating " + RubyLanguage.fileLine(getParent().getEncapsulatingSourceSection()) + " in " + getRootNode());
             }
 
-            resolved = resolver.get();
-            transferFlagsTo(resolved);
+            final RubyNode result = resolver.get();
+            transferFlagsTo(result);
 
-            replace(resolved, "lazy node resolved");
-            return resolved;
+            // publish
```

Comment:
I changed it to keep the LazyRubyNode so it's much simpler to make it thread-safe.